### ISSUE TITLE
Feature request: Secure Boot via shim support

### ISSUE THREAD

**Issue by geckolinux** [calamares/calamares#892]
Hi, I maintain a spin based on openSUSE Leap and Tumbleweed. The underlying OS and its native installer support Secure Boot out of the box via [shim](https://en.opensuse.org/openSUSE:UEFI):
> The default boot loader used by openSUSE 12.3 on UEFI systems is grub2. When in secure boot mode, an additional boot loader called 'shim' is used too. Instead of directly calling grub2 in that mode the firmware first loads 'shim'. 'shim' carries a signature by Microsoft in order to be recognized by the firmware. 'shim' in turn knows about the openSUSE certificate that was used to sign grub2. grub2 then is able to load linux kernels that are also signed by the openSUSE certificates. After loading the Linux kernel the scope of secure boot ends.

So the Calamares `bootloader.conf` should offer a third option of `efiBootLoader: "shim"`

For testing this feature, this blog describes how virt-manager and KVM can be used with a UEFI Secure Boot virtual firmware:
- https://nwrickert2.wordpress.com/2018/01/07/geckolinux-423-180105-a-review/
- https://nwrickert2.wordpress.com/2018/01/08/geckolinux-has-been-updated-to-180107/

Unfortunately with today's hardware that makes it so hard for users to install an alternative non-Windows OS, Secure Boot support is a necessary evil.

**Comment by kkofler** [comment]
As explained on IRC, this is also wanted in Fedora. Not only would it add "Secure Boot" support (which is of course a wanted feature to begin with), but it would also avoid having to drag in GRUB subpackages that are not installed by default on Fedora, which contain the modules necessary to run `grub2-install` on UEFI systems. Those subpackages are not installed by default because `shim` should be used instead (even on systems without "Secure Boot").

**Comment by adriaandegroot** [comment]
What files need to go where in that case? Is "shim" a file somewhere that needs copying? .. the first of the two links above says what the user **did** in the live system (run `shim-install`), but that didn't work. We'd need to know what commands to execute within the live system (or from outside) to install the right bits.

**Comment by bionel** [comment]
i have a strange feeling somebody does not understand how shim realy works.
quote from: https://github.com/rhboot/shim

> shim is a trivial EFI application that, when run, attempts to open and
execute another application. It will initially attempt to do this via the
standard EFI LoadImage() and StartImage() calls. If these fail (because secure
boot is enabled and the binary is not signed with an appropriate key, for
instance) it will then validate the binary against a built-in certificate. If
this succeeds and if the binary or signing key are not blacklisted then shim
will relocate and execute the binary.

so, skipping the 'trivial' word, which may not be the best choise in the description, the implementation on secure boot has to reside onto the "bootloader+kernel intended to boot" implementation (in a simplified sumary). there is nothing to do on the installer part, IMHO, since once you cannot boot the liveOS you cannot install it. and however shim it is loaded naively by the machine firmware, if it is properly set-up and placed into the right place, not by any 2nd (bootloader) or 3d (kernel) or ... (here the counting stops, i am too lazy to count all process involved until you get to calamares running))

more details about how shim and other (U)EFI realy works can be found at http://www.uefi.org/specifications

**Comment by bionel** [comment]
@geckolinux 
since you are mentioning you are a maintainer of an OpenSUsE spin i will strongly suggest to read about how they did it:
https://en.opensuse.org/openSUSE:UEFI#How_to_enable_or_disable_Secure_Boot_support

their approach is getting an M$ signed binary to enable 'we can boot' however AFAIK they still dont have the full chain with own_signed_grub_kernel_modules etc. not a reliable solution, just a hack.

A better approach will be  to give to the user the power to sign his own binaries:
https://www.rodsbooks.com/efi-bootloaders/secureboot.html

**Comment by kkofler** [comment]
> there is nothing to do on the installer part, IMHO, since once you cannot boot the liveOS you cannot install it.

This is already your first incorrect assertion. Fedora live images, and even third-party Fedora Remix live images, boot just fine on "Secure Boot" machines.

> and however shim it is loaded natively by the machine firmware, if it is properly set-up and placed into the right place, not by any 2nd (bootloader) or 3d (kernel) or ...

`shim` takes the place of the bootloader, so instead of installing the bootloader, we have to install `shim`. The live images include a copy of `shim` signed with the M$ key (signed copy of `shim` which is available from the Fedora repository) in their EFI system partition, the same copy has to be installed to the EFI system partition on the hard disk.

> their approach is getting an M$ signed binary to enable 'we can boot' however AFAIK they still dont have the full chain with own_signed_grub_kernel_modules etc. not a reliable solution, just a hack.

And this feature requests installing that signed binary as a bootloader (which is set up to chainload the GRUB bootloader without having to explicitly install GRUB anywhere) so that the installed distribution will boot. This is exactly what the native installer (e.g., Anaconda on Fedora) does.

> A better approach will be to give to the user the power to sign his own binaries:

`shim` already allows it, you can enroll your own keys without going through the firmware's interface for it (which might not even exist), but you have to interactively confirm the key installation (because anything non-interactive would never pass M$'s signature requirements because it could easily be abused by malware). The link you posted already explains that here:
https://www.rodsbooks.com/efi-bootloaders/secureboot.html#signing

There are more detailed and more official explanations, e.g.:
* https://drivers.suse.com/doc/Usage/Secure_Boot_Certificate.html#secure-boot-certificate
* https://fedoraproject.org/wiki/BuildingUpstreamKernel#Sign_the_kernel_for_Secure_Boot

So support for `shim` actually already automatically gives the user that power. We cannot automatically enroll a signing key for the user because if Calamares can do it, so can malware, and thus `shim` requires user interaction to enroll a key. And the option of enrolling keys directly in the firmware, if it even allows it, is something we cannot control or use at all, it will always be there, or not. So there is nothing more we can do in Calamares, `shim` support already gives you what we can support.

And the option to disable "Secure Boot" altogether is not affected by this feature at all. `shim` works fine with "Secure Boot" disabled and automatically disables its own signature checks in that case.

**Comment by kkofler** [comment]
The way Anaconda installs `shim` is using `efibootmgr` with a bunch of options, see the `EFIBase` and `EFIGRUB` classes:
* https://github.com/rhinstaller/anaconda/blob/d9495cc18766f2717178f8ea8a75d6a352346f80/pyanaconda/bootloader.py#L1675
* https://github.com/rhinstaller/anaconda/blob/d9495cc18766f2717178f8ea8a75d6a352346f80/pyanaconda/bootloader.py#L1822

**Comment by kkofler** [comment]
And openSUSE ships a `shim-install` script: https://build.opensuse.org/package/view_file/devel:openSUSE:Factory/shim/shim-install?expand=1 which is basically a wrapper around `efibootmgr` passing the right options, so calling `efibootmgr` directly should work there too.

**Comment by bionel** [comment]
> This is already your first incorrect assertion. Fedora live images, and even third-party Fedora Remix live images, boot just fine on "Secure Boot" machines.

Kindly read it properly. Fedora and so on *can* boot in secure boot, therefore can be installed without big fuss.

> shim takes the place of the bootloader ...

Not quiet correct. Kindly read again the provided UEFI spec and the shim documentation. The only thing shim is doing is to intermediate between UEFI and the mechanism to boot (bootloader, loading kernel directly, systemd and so on) in case the next stage binary is not signed. It is not a bootloader by definition, therefore cannot take his place. It may be better to name it 'elf loader'

> So support for shim actually already automatically gives the user that power ... So there is nothing more we can do in Calamares, shim support already gives you what we can support

This is what I said from the beginning :) The implementation (through shim or preLoader or any other mechanism) should be on the distribution architecture side (delivering already signed binaries, give the user the choice to sign/enroll his own etc).

**Comment by kkofler** [comment]
We seem to be talking past each other. I never claimed that `shim` is a bootloader, all I'm saying is that it is installed to the same location as the bootloader and is treated like a bootloader by the UEFI firmware. But then, it will only signature-check and then load the actual bootloader and not perform any bootloader tasks itself.

And what we are asking from Calamares is to install `shim` to the EFI System Partition (and GRUB to somewhere `shim` will find it, if necessary on the distribution – as far as I know, the Fedora `grub2-efi` package already installs the needed files to the right place) instead of installing GRUB there. For that, we want an `efiBootLoader: "shim"` option (in addition to the existing `"grub"` and `"systemd-boot"` options that will not be touched in any way). Of course, that option will only work on distributions that are set up for `shim` to begin with, I never claimed otherwise. But other distributions (which do not support "Secure Boot") will still be able to use the existing 2 options.

**Comment by adriaandegroot** [comment]
```
➜ cat before_bootloader_context.conf 
# Make sure the correct bootloader package is installed for efi.
# Also pull in the shim so secureboot has a chance at working.
---
firmwareType:
    efi:
    - command: apt install -y --no-upgrade -o Acquire::gpgv::Options::=--ignore-time-conflict grub-efi-amd64-signed
      timeout: 300
    - command: apt install -y --no-upgrade -o Acquire::gpgv::Options::=--ignore-time-conflict shim-signed
      timeout: 300

```

That's a contextualprocess approach to the whole thing, for Neon.

**Comment by geckolinux** [comment]
Hi @adriaandegroot , thanks for adding this to the next milestone. I'm excited to see what comes of it.

**Comment by kkofler** [comment]
This feature was implemented in `master` with commit 5643ee4b8f17826f0067a217e4e99eddee05f616. (The commit message says "Merge branch 'master' into secureboot", but `secureboot` was then fast-forward-merged back into `master`.)

**Comment by geckolinux** [comment]
@Kevin Kofler: Nice! Any tricks on setting it up? Calamares config file changes?

**Comment by kkofler** [comment]
@geckolinux:
> Any tricks on setting it up? Calamares config file changes?

In `bootloader.conf`, you have to change `efiBootLoader: "grub"` to `efiBootLoader: "sb-shim"` (the actual bootloader will still be GRUB2-EFI, but it will be loaded through `shim` rather than directly), and to ensure that the `grubProbe` and `efiBootMgr` settings are correct for your distribution.

(The same goes for all distributions that want to use this feature.)

**Comment by geckolinux** [comment]
Hi everyone, sorry I don't quite understand how the commits and branches correlate to actual releases. Is commit 5643ee4 included in Calamares 3.2.1?
